File: env_manager.py
# env_manager.py - 环境变量管理和验证系统
import os
import re
from typing import Dict, List, Optional, Tuple, Any
from dataclasses import dataclass
from enum import Enum
import json


class ConfigStatus(Enum):
    """配置状态枚举"""
    VALID = "valid"
    MISSING = "missing"
    INVALID = "invalid"
    WARNING = "warning"


@dataclass
class ConfigItem:
    """配置项数据类"""
    name: str
    value: Optional[str]
    status: ConfigStatus
    message: str
    required: bool = True


@dataclass
class ConfigReport:
    """配置状态报告"""
    items: List[ConfigItem]
    overall_status: ConfigStatus
    summary: str
    
    def to_dict(self) -> Dict[str, Any]:
        """转换为字典格式"""
        return {
            "items": [
                {
                    "name": item.name,
                    "value": item.value,
                    "status": item.status.value,
                    "message": item.message,
                    "required": item.required
                }
                for item in self.items
            ],
            "overall_status": self.overall_status.value,
            "summary": self.summary
        }


class EnvironmentManager:
    """环境变量管理器"""
    
    # 必需的环境变量配置
    REQUIRED_VARS = {
        "SECRET_KEY": {
            "description": "JWT密钥",
            "validator": lambda x: len(x) >= 32,
            "error_msg": "SECRET_KEY必须至少32个字符"
        },
        "CORS_ORIGINS": {
            "description": "CORS允许的源",
            "validator": lambda x: EnvironmentManager._validate_cors_origins(x),
            "error_msg": "CORS_ORIGINS格式无效，应为逗号分隔的URL列表"
        }
    }
    
    # 可选的环境变量配置
    OPTIONAL_VARS = {
        "DATABASE_URL": {
            "description": "数据库连接URL",
            "validator": lambda x: x.startswith(("sqlite://", "postgresql://", "mysql://")),
            "error_msg": "DATABASE_URL格式无效",
            "default": "sqlite:///./recipes.db"
        },
        "RAILWAY_STATIC_URL": {
            "description": "Railway静态URL",
            "validator": lambda x: x.startswith("https://"),
            "error_msg": "RAILWAY_STATIC_URL必须是HTTPS URL",
            "default": None
        }
    }
    
    @staticmethod
    def _validate_cors_origins(origins_str: str) -> bool:
        """验证CORS源格式"""
        if not origins_str:
            return False
        
        origins = [o.strip() for o in origins_str.split(",") if o.strip()]
        if not origins:
            return False
        
        # 允许 * 或有效的URL
        for origin in origins:
            if origin == "*":
                continue
            if not re.match(r'^https?://[a-zA-Z0-9.-]+(?::\d+)?$', origin):
                return False
        
        return True
    
    @staticmethod
    def _validate_url(url: str) -> bool:
        """验证URL格式"""
        url_pattern = re.compile(
            r'^https?://'  # http:// 或 https://
            r'(?:(?:[A-Z0-9](?:[A-Z0-9-]{0,61}[A-Z0-9])?\.)+[A-Z]{2,6}\.?|'  # 域名
            r'localhost|'  # localhost
            r'\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})'  # IP地址
            r'(?::\d+)?'  # 可选端口
            r'(?:/?|[/?]\S+)$', re.IGNORECASE)
        return url_pattern.match(url) is not None
    
    def validate_environment_variables(self) -> ConfigReport:
        """验证所有环境变量"""
        items = []
        has_errors = False
        has_warnings = False
        
        # 检查必需变量
        for var_name, config in self.REQUIRED_VARS.items():
            value = os.getenv(var_name)
            
            if value is None:
                items.append(ConfigItem(
                    name=var_name,
                    value=None,
                    status=ConfigStatus.MISSING,
                    message=f"缺少必需的环境变量: {config['description']}",
                    required=True
                ))
                has_errors = True
            elif not config["validator"](value):
                items.append(ConfigItem(
                    name=var_name,
                    value=value,
                    status=ConfigStatus.INVALID,
                    message=config["error_msg"],
                    required=True
                ))
                has_errors = True
            else:
                items.append(ConfigItem(
                    name=var_name,
                    value=value,
                    status=ConfigStatus.VALID,
                    message=f"✓ {config['description']}配置正确",
                    required=True
                ))
        
        # 检查可选变量
        for var_name, config in self.OPTIONAL_VARS.items():
            value = os.getenv(var_name)
            
            if value is None:
                default_value = config.get("default")
                if default_value:
                    items.append(ConfigItem(
                        name=var_name,
                        value=default_value,
                        status=ConfigStatus.WARNING,
                        message=f"使用默认值: {default_value}",
                        required=False
                    ))
                    has_warnings = True
                else:
                    items.append(ConfigItem(
                        name=var_name,
                        value=None,
                        status=ConfigStatus.WARNING,
                        message=f"可选变量未设置: {config['description']}",
                        required=False
                    ))
                    has_warnings = True
            elif not config["validator"](value):
                items.append(ConfigItem(
                    name=var_name,
                    value=value,
                    status=ConfigStatus.INVALID,
                    message=config["error_msg"],
                    required=False
                ))
                has_warnings = True
            else:
                items.append(ConfigItem(
                    name=var_name,
                    value=value,
                    status=ConfigStatus.VALID,
                    message=f"✓ {config['description']}配置正确",
                    required=False
                ))
        
        # 确定整体状态
        if has_errors:
            overall_status = ConfigStatus.INVALID
            summary = "配置验证失败：存在必需变量缺失或无效"
        elif has_warnings:
            overall_status = ConfigStatus.WARNING
            summary = "配置基本正确，但存在警告项"
        else:
            overall_status = ConfigStatus.VALID
            summary = "所有配置项验证通过"
        
        return ConfigReport(
            items=items,
            overall_status=overall_status,
            summary=summary
        )
    
    def get_api_base_url(self) -> str:
        """获取API基础URL"""
        # 优先使用Railway提供的URL
        railway_url = os.getenv("RAILWAY_STATIC_URL")
        if railway_url:
            return railway_url
        
        # 回退到本地开发URL
        return "http://localhost:8000"
    
    def get_cors_origins(self) -> List[str]:
        """获取CORS允许的源"""
        origins_raw = os.getenv("CORS_ORIGINS", "http://localhost:5173")
        origins = [o.strip() for o in origins_raw.split(",") if o.strip()]
        return origins
    
    def setup_environment_defaults(self) -> None:
        """设置环境变量默认值（如果未设置）"""
        defaults = {
            "SECRET_KEY": "09d25e094faa6ca2556c818166b7a9563b93f7099f6f0f4caa6cf63b88e8d3e7",
            "CORS_ORIGINS": "http://localhost:5173",
            "DATABASE_URL": "sqlite:///./recipes.db"
        }
        
        for key, default_value in defaults.items():
            if not os.getenv(key):
                os.environ[key] = default_value
    
    def generate_config_report_json(self) -> str:
        """生成JSON格式的配置报告"""
        report = self.validate_environment_variables()
        return json.dumps(report.to_dict(), ensure_ascii=False, indent=2)
    
    def print_config_status(self) -> None:
        """打印配置状态到控制台"""
        report = self.validate_environment_variables()
        
        print("=" * 50)
        print("环境变量配置状态报告")
        print("=" * 50)
        print(f"整体状态: {report.overall_status.value.upper()}")
        print(f"摘要: {report.summary}")
        print()
        
        # 按状态分组显示
        for status in [ConfigStatus.INVALID, ConfigStatus.MISSING, ConfigStatus.WARNING, ConfigStatus.VALID]:
            status_items = [item for item in report.items if item.status == status]
            if status_items:
                print(f"{status.value.upper()} 项目:")
                for item in status_items:
                    required_mark = "[必需]" if item.required else "[可选]"
                    print(f"  {required_mark} {item.name}: {item.message}")
                print()


# 全局环境管理器实例
env_manager = EnvironmentManager()


def validate_environment() -> ConfigReport:
    """验证环境变量的便捷函数"""
    return env_manager.validate_environment_variables()


def get_config_status() -> Dict[str, Any]:
    """获取配置状态的便捷函数"""
    report = env_manager.validate_environment_variables()
    return report.to_dict()


if __name__ == "__main__":
    # 测试环境变量管理器
    manager = EnvironmentManager()
    manager.print_config_status()

File: frontend/src/envManager.ts
// envManager.ts - 前端环境变量管理和验证系统
export const ConfigStatus = {
  VALID: 'valid',
  MISSING: 'missing',
  INVALID: 'invalid',
  WARNING: 'warning'
} as const;

export type ConfigStatus = typeof ConfigStatus[keyof typeof ConfigStatus];

export interface ConfigItem {
  name: string;
  value: string | null;
  status: ConfigStatus;
  message: string;
  required: boolean;
}

export interface ConfigReport {
  items: ConfigItem[];
  overallStatus: ConfigStatus;
  summary: string;
}

export interface EnvironmentConfig {
  apiBaseUrl: string;
  isProduction: boolean;
  buildTime: string;
}

class FrontendEnvironmentManager {
  // 必需的环境变量配置
  private readonly REQUIRED_VARS = {
    VITE_API_BASE_URL: {
      description: 'API基础URL',
      validator: (value: string) => this.validateUrl(value),
      errorMsg: 'VITE_API_BASE_URL必须是有效的URL'
    }
  };

  // 可选的环境变量配置
  private readonly OPTIONAL_VARS = {
    VITE_APP_TITLE: {
      description: '应用标题',
      validator: (value: string) => value.length > 0,
      errorMsg: 'VITE_APP_TITLE不能为空',
      default: 'AI食谱推荐'
    },
    VITE_DEBUG: {
      description: '调试模式',
      validator: (value: string) => ['true', 'false'].includes(value.toLowerCase()),
      errorMsg: 'VITE_DEBUG必须是true或false',
      default: 'false'
    }
  };

  /**
   * 验证URL格式
   */
  private validateUrl(url: string): boolean {
    try {
      const urlObj = new URL(url);
      return ['http:', 'https:'].includes(urlObj.protocol);
    } catch {
      return false;
    }
  }

  /**
   * 获取环境变量值
   */
  private getEnvVar(name: string): string | null {
    // 在构建时，Vite会将环境变量注入到import.meta.env中
    const value = (import.meta.env as any)[name];
    return value || null;
  }

  /**
   * 验证所有环境变量
   */
  public validateEnvironmentVariables(): ConfigReport {
    const items: ConfigItem[] = [];
    let hasErrors = false;
    let hasWarnings = false;

    // 检查必需变量
    for (const [varName, config] of Object.entries(this.REQUIRED_VARS)) {
      const value = this.getEnvVar(varName);

      if (value === null) {
        items.push({
          name: varName,
          value: null,
          status: ConfigStatus.MISSING,
          message: `缺少必需的环境变量: ${config.description}`,
          required: true
        });
        hasErrors = true;
      } else if (!config.validator(value)) {
        items.push({
          name: varName,
          value: value,
          status: ConfigStatus.INVALID,
          message: config.errorMsg,
          required: true
        });
        hasErrors = true;
      } else {
        items.push({
          name: varName,
          value: value,
          status: ConfigStatus.VALID,
          message: `✓ ${config.description}配置正确`,
          required: true
        });
      }
    }

    // 检查可选变量
    for (const [varName, config] of Object.entries(this.OPTIONAL_VARS)) {
      const value = this.getEnvVar(varName);

      if (value === null) {
        const defaultValue = config.default;
        items.push({
          name: varName,
          value: defaultValue,
          status: ConfigStatus.WARNING,
          message: `使用默认值: ${defaultValue}`,
          required: false
        });
        hasWarnings = true;
      } else if (!config.validator(value)) {
        items.push({
          name: varName,
          value: value,
          status: ConfigStatus.INVALID,
          message: config.errorMsg,
          required: false
        });
        hasWarnings = true;
      } else {
        items.push({
          name: varName,
          value: value,
          status: ConfigStatus.VALID,
          message: `✓ ${config.description}配置正确`,
          required: false
        });
      }
    }

    // 确定整体状态
    let overallStatus: ConfigStatus;
    let summary: string;

    if (hasErrors) {
      overallStatus = ConfigStatus.INVALID;
      summary = '配置验证失败：存在必需变量缺失或无效';
    } else if (hasWarnings) {
      overallStatus = ConfigStatus.WARNING;
      summary = '配置基本正确，但存在警告项';
    } else {
      overallStatus = ConfigStatus.VALID;
      summary = '所有配置项验证通过';
    }

    return {
      items,
      overallStatus,
      summary
    };
  }

  /**
   * 获取API基础URL
   */
  public getApiBaseUrl(): string {
    const configuredUrl = this.getEnvVar('VITE_API_BASE_URL');
    
    // 如果配置了环境变量且有效，使用它
    if (configuredUrl && this.validateUrl(configuredUrl)) {
      return configuredUrl;
    }

    // 在生产环境中，尝试自动检测
    if (import.meta.env.PROD) {
      // 如果是在Railway等平台部署，尝试使用相对路径
      const currentOrigin = window.location.origin;
      
      // 检查是否有Railway域名模式
      if (currentOrigin.includes('railway.app')) {
        // 尝试多种Railway后端URL模式
        const possibleBackendUrls = [
          currentOrigin.replace('-frontend', '-backend'),
          currentOrigin.replace('frontend-', 'backend-'),
          currentOrigin.replace('frontend.', 'backend.'),
          // 如果前端域名包含项目名，尝试构建后端URL
          currentOrigin.replace(/frontend/g, 'backend')
        ];
        
        // 返回第一个可能的URL（实际使用时会通过连接测试验证）
        return possibleBackendUrls[0];
      }
      
      // 默认假设后端在同一域名的不同端口或路径
      return currentOrigin;
    }

    // 开发环境默认值
    return 'http://localhost:8000';
  }

  /**
   * 获取构建时信息
   */
  public getBuildInfo(): { buildTime: string; version: string; environment: string } {
    return {
      buildTime: this.getEnvVar('VITE_BUILD_TIME') || 'unknown',
      version: (globalThis as any).__VERSION__ || '1.0.0',
      environment: import.meta.env.MODE || 'development'
    };
  }

  /**
   * 验证构建时配置
   */
  public validateBuildConfiguration(): { isValid: boolean; errors: string[]; warnings: string[] } {
    const errors: string[] = [];
    const warnings: string[] = [];

    // 检查必需的构建时变量
    const apiUrl = this.getEnvVar('VITE_API_BASE_URL');
    if (!apiUrl) {
      errors.push('VITE_API_BASE_URL 未在构建时设置');
    } else if (!this.validateUrl(apiUrl)) {
      errors.push(`VITE_API_BASE_URL 格式无效: ${apiUrl}`);
    }

    // 检查构建时间戳
    const buildTime = this.getEnvVar('VITE_BUILD_TIME');
    if (!buildTime) {
      warnings.push('VITE_BUILD_TIME 未设置，可能影响缓存管理');
    }

    // 检查应用标题
    const appTitle = this.getEnvVar('VITE_APP_TITLE');
    if (!appTitle) {
      warnings.push('VITE_APP_TITLE 未设置，将使用默认标题');
    }

    return {
      isValid: errors.length === 0,
      errors,
      warnings
    };
  }

  /**
   * 获取环境配置
   */
  public getEnvironmentConfig(): EnvironmentConfig {
    return {
      apiBaseUrl: this.getApiBaseUrl(),
      isProduction: import.meta.env.PROD,
      buildTime: import.meta.env.VITE_BUILD_TIME || new Date().toISOString()
    };
  }

  /**
   * 打印配置状态到控制台
   */
  public printConfigStatus(): void {
    const report = this.validateEnvironmentVariables();
    
    console.group('🔧 环境变量配置状态报告');
    console.log(`整体状态: ${report.overallStatus.toUpperCase()}`);
    console.log(`摘要: ${report.summary}`);
    console.log('');

    // 按状态分组显示
    const statusOrder = [ConfigStatus.INVALID, ConfigStatus.MISSING, ConfigStatus.WARNING, ConfigStatus.VALID];
    
    for (const status of statusOrder) {
      const statusItems = report.items.filter(item => item.status === status);
      if (statusItems.length > 0) {
        console.group(`${status.toUpperCase()} 项目:`);
        for (const item of statusItems) {
          const requiredMark = item.required ? '[必需]' : '[可选]';
          console.log(`${requiredMark} ${item.name}: ${item.message}`);
        }
        console.groupEnd();
      }
    }
    
    console.groupEnd();
  }

  /**
   * 生成配置报告的JSON字符串
   */
  public generateConfigReportJson(): string {
    const report = this.validateEnvironmentVariables();
    return JSON.stringify(report, null, 2);
  }

  /**
   * 检查配置是否就绪
   */
  public isConfigurationReady(): boolean {
    const report = this.validateEnvironmentVariables();
    return report.overallStatus !== ConfigStatus.INVALID;
  }
}

// 创建全局实例
export const frontendEnvManager = new FrontendEnvironmentManager();

// 便捷函数
export function validateFrontendEnvironment(): ConfigReport {
  return frontendEnvManager.validateEnvironmentVariables();
}

export function getFrontendConfig(): EnvironmentConfig {
  return frontendEnvManager.getEnvironmentConfig();
}

export function getApiBaseUrl(): string {
  return frontendEnvManager.getApiBaseUrl();
}

// 在开发模式下自动打印配置状态
if (import.meta.env.DEV) {
  frontendEnvManager.printConfigStatus();
}